resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala

```diff
@@ -525,9 +534,11 @@ private[yarn] class YarnAllocator(
               } catch {
                 case NonFatal(e) =>
                   logError(s"Failed to launch executor $executorId on container $containerId", e)
-                  // Assigned container should be released immediately to avoid unnecessary resource
-                  // occupation.
+                  // Assigned container should be released immediately
+                  // to avoid unnecessary resource occupation.
                   amClient.releaseAssignedContainer(containerId)
+              } finally {
+                numExecutorsStarting.decrementAndGet()
```

Comment:
not sure I follow your reference between starting and running.  I was just saying in the failure case it doesn't matter because you aren't going to overcount.

If we don't decrement the starting inside of updateInternalState we have the possibility to over count because running and starting aren't incremented/decremented within the synchronize block.  We don't want to do that.